software/base/src/main/java/org/apache/brooklyn/entity/software/base/lifecycle/MachineLifecycleEffectorTasks.java

```diff
@@ -709,7 +712,26 @@ protected void doStop(ConfigBag parameters, Callable<StopMachineDetails<Integer>
 
         DynamicTasks.queue("pre-stop", new PreStopCustomTask());
 
+        // BROOKLYN-263:
+        // With this change the stop effector will wait for Location to provision so it can stop it.
+        // It will not retry if non-jclouds location is used in the entity.
         Maybe<MachineLocation> machine = Machines.findUniqueMachineLocation(entity().getLocations());
+        if (Boolean.TRUE.equals(entity().sensors().get(Attributes.JCLOUDS_PROVISIONING_STARTED))
+                && machine.isAbsent()) {
+            entity().sensors().set(Attributes.JCLOUDS_PROVISIONING_STARTED, null);
+            Repeater.create("Wait for a machine to appear")
+                    .until(new Callable<Boolean>() {
+                        @Override
+                        public Boolean call() throws Exception {
+                            return Machines.findUniqueMachineLocation(entity().getLocations()).isPresent();
+                        }
+                    })
+                    .every(Duration.THIRTY_SECONDS)
```

Comment:
And to be able to test it fast (in unit tests, rather than live tests), we'd want to set this very low. But let's not worry about that quite yet.